examples/opt_serving/interactive_hosted.py

```diff
@@ -105,13 +116,80 @@ def batching_loop(timeout=TIMEOUT_MS, max_tokens=MAX_BATCH_TOKENS, max_bs=MAX_BS
+                        "min_tokens": [],
+                        "max_tokens": [],
+                    }
+                    ro = work_item.data
+                    request_object["inputs"].append(ro["input"])
+                    request_object["min_tokens"].append(ro.get("min_tokens", 0))
+                    request_object["max_tokens"].append(
+                        ro.get("max_tokens", MAX_SEQ_LEN))
+                    cache_ids.append(ro.get("cache_id", None))
+                    # assumption: everyone has the same remaining args
+                    for key in [
+                            "temperature",
+                            "top_p",
+                            "top_k",
+                            "n",
+                            "best_of",
+                            "echo",
+                            "logprobs",
+                            "stop",
+                    ]:
+                        if key in ro:
+                            request_object[key] = ro[key]
+                    # do the actual generations
+                    request_object["seed"] = random.randint(1, 20000)
+                    outputs = generator(request_object['inputs'], cache_ids, pasts=PAST_CACHE)
+                    generations = []
+                    for cache_id, output in zip(cache_ids, outputs):
+                        PAST_CACHE[cache_id] = (time.time(), output.past_key_values) # add to or update the cache with newly computed values
+                        if len(PAST_CACHE) > PAST_CACHE_SIZE_LIMIT: # delete oldest key in cache if cache too big
+                            oldest_key = min(list(PAST_CACHE.keys()), key=lambda k: PAST_CACHE[k][0])
+                            del PAST_CACHE[oldest_key]
+                        logprobs = torch.log_softmax(output.logits[0, 0], dim=-1)
+                        sorted_logprobs, sorted_indices = torch.sort(logprobs, descending=True)
+                        cumulative_probs = torch.cumsum(torch.softmax(sorted_logprobs, dim=-1), dim=-1)
+                        num_to_return = 1 + (cumulative_probs < request_object['top_p']).sum().item()
+                        num_to_return = min(num_to_return, request_object["top_k"]) # return at most top_k tokens, e.g. if network limited
+                        generations.append({'logprobs': sorted_logprobs[:num_to_return].tolist(), 'indices': sorted_indices[:num_to_return].tolist()})
+                    # broadcast them back
+                    for work_item, gen in zip(logprobs_batch, generations):
+                        work_item.return_queue.put((work_item.uid, gen))
+
+                    logprobs_batch.clear()
+        
+        # clear old keys in PAST_CACHE which haven't been used recently
+        existing_cache_ids = list(PAST_CACHE.keys())
+        for cache_id in existing_cache_ids:
+            if time.time() - PAST_CACHE[cache_id][0] > PAST_CACHE_TIMEOUT:
+                del PAST_CACHE[cache_id]
```

Comment:
If I understand correctly, PAST_CACHE will be a huge cache that occupies a large portion of GPU memory. If we only use the TIMEOUT mechanism, will it make our original service OOM due to the cache in GPU?

Reply:
I also wrote a mechanism to remove the oldest item once it exceeds a specified maximum size, regardless of timeout, based on `PAST_CACHE_SIZE_LIMIT`. So the cache should be limited to some maximum size. 

Additionally, it might be possible to cache only the relevant keys instead of the full padded batch of size 16 by manipulating the past_key_values, but I wasn't sure if constantly reconstructing the padded keys on GPU would introduce a lot of overhead so I didn't do it for now. 

I'm not sure how much of a limitation the memory would be on your server; happy to modify the numbers/implementation based on what you think the right tradeoff is.